kedro_azureml/runner.py

```diff
@@ -33,25 +23,20 @@ def run(
         hook_manager: PluginManager = None,
         session_id: str = None,
     ) -> Dict[str, Any]:
-        unsatisfied = pipeline.inputs() - set(catalog.list())
-        for ds_name in unsatisfied:
-            catalog = catalog.shallow_copy()
-            catalog.add(ds_name, self.create_default_data_set(ds_name))
+        catalog = catalog.shallow_copy()
+
+        # Loop over input and output datasets in arguments to set their paths
+        for ds_name, ds_path in self.data_paths.items():
+            if ds_name in catalog.list():
+                ds = catalog._get_dataset(ds_name)
+                ds._filepath = Path(ds_path) / Path(ds._filepath).name
+                # Disable versioning when passing data between nodes
+                ds._version = None
+                catalog.add(ds_name, ds, replace=True)
+            else:
+                catalog.add(ds_name, self.create_default_data_set(ds_name))
 
         return super().run(pipeline, catalog, hook_manager, session_id)
 
     def create_default_data_set(self, ds_name: str) -> AbstractDataSet:
-        # TODO: handle credentials better (probably with built-in Kedro credentials
-        #  via ConfigLoader (but it's not available here...)
-        dataset_cls = KedroAzureRunnerDataset
-        if is_distributed_environment():
-            logger.info("Using distributed dataset class as a default")
-            dataset_cls = KedroAzureRunnerDistributedDataset
-
-        return dataset_cls(
-            self.runner_config.temporary_storage.account_name,
-            self.runner_config.temporary_storage.container,
-            self.runner_config.storage_account_key,
-            ds_name,
-            self.runner_config.run_id,
-        )
+        return PickleDataSet(str(Path(self.data_paths[ds_name]) / f"{ds_name}.pickle"))
```

Comment:
After our discussion I've thought about it and I understand the desire to have "preview-able" intermediate datasets for Data Scientists to explore from within the Azure ML Workspace.
The way to do this that will be the most compatible with Kedro way of handling the catalog would be to redesign this feature to behave more or less like this:

1. Create configuration entry for `azureml.yml` that will enable this feature (disabled by default)
2. If the feature is enabled, then any dataset that is NOT defined in the catalog will use new dataset implementation (let's call it `AzureMLPreviewableDataSet` for the sake of this comment):
```python
def create_default_data_set(self, ds_name: str) -> AbstractDataSet:
        if <feature is enabled>:
            return AzureMLPreviewableDataSet(<some params>)
        else:
               <old behaviour>
```
3. In the `AzureMLPreviewableDataSet` during `_save`:
* Check data type that is being saved. There are not that many human-readable data types from the DS perspective, I guess it will be: pandas dataframes, pure text, images (maybe sth else, this can always be extended in the future).
* For every data type, choose one way of saving, so in the `_save` there will be something like:

```python
if isinstance(data, pd.DataFrame):
    # save as CSV to the data_path from Azure
elif isinstance(data, PIL.Image):
     # save as image to the data_path from Azure
elif isinstance(data, str):
     # save as text to the data_path from Azure
else:
     # save as a compressed cloudpickle - the current behavior from KedroAzureRunnerDataset>
```

* During load do the reverse (will require some try-catching, because the loading might likely fail) but it's definitely doable.


That way of implementing it satisfies: reverse compatibility, datasets will be previewable for data scientists, intermediate data will be saved using native-AzureML data passing.

Reply:
As discussed, trying to infer the dataset type is not really desirable. The current implementation allows to change the format from the default pickle by specifying it in the catalog as a `AzureMLFolderDataset`.